### DISCUSSION TITLE
Lexus support

### DISCUSSION THREAD

**Discussion by Gorgounet** [DurgNomis-drol/ha_toyota#218] [Ideas]
Hello guys,

I was wondering if an integration for Lexus was in the pipe as they are theoretically based on the same system the main difference seems to be the layer when using the website or the app which is changing.

Cheers

**Comment by CM000n** [comment]
In principle, Lexus should use the same API. If you have a Lexus, you can try to see if you can log in. That would be interesting.
A dedicated Lib or custoim component for Lexus is currently not planned, at least not from my side.

**Comment by Gorgounet** [comment]
That is a very good point and I tried, however even if my Lexus account is also a Toyota account as you can log on both apps/websites with the same email address, when trying to login through home assistant, I am getting an Unexpected error message whatever I try to do

**Reply by CM000n** [reply]
Could you please open an issue in the mytoyota Python Lib: https://github.com/DurgNomis-drol/mytoyota/issues
Can you please execute the `simple_client_example.py` there and post the error output.

First of all, we need to have a basic look at whether and how a Lexus connection works with the mytoyota Python package.

Can you also use the MyToyota app for your Lexus, or is that a standalone app? Lexus may also be on an older/different API version here

**Reply by CM000n** [reply]
It looks like Lexus has recently also changed its apps from the [Lexus Link app](https://play.google.com/store/apps/details?id=app.mylexus.lexus.com.mylexus) to the [Lexus Link+](https://play.google.com/store/apps/details?id=com.lexus.oneapp.eu) app.
Which of the two apps do you use @Gorgounet?

**Reply by ellvisss** [reply]
I have a Lexus and I am using Lexus Link+
I have tried to execute the `simple_client_example.py`:

> ...\mytoyota-master\simple_client_example.py:86: DeprecationWarning: There is no current event loop
> loop = asyncio.get_event_loop()
> Logging in...
> Retrieving cars...
> [Finished in 2.7s]

I do not have any cars in MyToyota and I cannot add a Lexus car. I have to use Lexus Link+. 

I also found this and I have tried it:
```
import requests
url = "https://ssoms.toyota-europe.com/authenticate"
payload = {'username': 'xx',
           'password': 'yy'}
files = []
headers = {
    'X-TME-BRAND': 'LEXUS',
    'X-TME-LC': 'ro-ro',
    'Accept': 'application / json, text / plain, * / *',
    'Sec-Fetch-Dest': 'empty'
}
response = requests.request(
    "POST", url, headers=headers, data=payload, files=files)
print(response.text)
```
and I have modified  `X-TME-BRAND` from `TOYOTA` to `LEXUS`. I have the same response that I do not have a car (`"extras":{"hasPurchasedCars":false}`). If I will modify to something else (AUDI for example), I get this: 

> Brand is either not supported, missing or malformed
> [Finished in 804ms

Please let me know if you need additional information.

**Reply by CM000n** [reply]
Thanks for testing it out. You could also try the Endpoint Tester from this branch to see if you can get even more detailed information directly from the endpoints.
https://github.com/DurgNomis-drol/mytoyota/pull/309

Just download the latest version of the branch and run `python endpoint_tester.py all_get.txt > responses.log`.
After that you can look at the response.log file.

**Reply by ellvisss** [reply]
Strangely, with `simple_client_example.py` even if I change the password in `credentials.json`, I get the same output. If I change the email without `@` or remove the password, I get an error.

I also tried `endpoint_tester.py` and I have this (with good or wrong password):
```
...\mytoyota-master\endpoint_tester.py:78: DeprecationWarning: There is no current event loop
  loop = asyncio.get_event_loop()
['...\\mytoyota-master\\endpoint_tester.py']
Traceback (most recent call last):
  File "...\mytoyota-master\endpoint_tester.py", line 79, in <module>
    loop.run_until_complete(main())
  File "C:\Users\UserName\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "...\mytoyota-master\endpoint_tester.py", line 55, in main
    assert len(vehicles) > 0
           ^^^^^^^^^^^^^^^^^
AssertionError
```

Edit: It seems that only the wrong user is giving an error:
I set username to `1@1.com` and I got this:

> ...\mytoyota-master\simple_client_example.py:86: DeprecationWarning: There is no current event loop
>   loop = asyncio.get_event_loop()
> Logging in...
> Traceback (most recent call last):
>   File "...\mytoyota-master\simple_client_example.py", line 87, in <module>
>     loop.run_until_complete(get_information())
>   File "C:\Users\UserName\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 653, in run_until_complete
>     return future.result()
>            ^^^^^^^^^^^^^^^
>   File "...\mytoyota-master\simple_client_example.py", line 47, in get_information
>     await client.login()
>   File "...\mytoyota-master\mytoyota\client.py", line 59, in login
>     await self._api.controller.login()
>   File "...\mytoyota-master\mytoyota\controller.py", line 68, in login
>     await self._update_token()
>   File "...\mytoyota-master\mytoyota\controller.py", line 80, in _update_token
>     await self._authenticate()
>   File "...\mytoyota-master\mytoyota\controller.py", line 101, in _authenticate
>     raise ToyotaInvalidUsernameError(
> mytoyota.exceptions.ToyotaInvalidUsernameError: Authentication Failed. User Not Found.

With the good username and wrong password, I get the same ouput:

> ...\mytoyota-master\simple_client_example.py:86: DeprecationWarning: There is no current event loop
> loop = asyncio.get_event_loop()
> Logging in...
> Retrieving cars...
> [Finished in 1.3s]

**Reply by Gorgounet** [reply]
> It looks like Lexus has recently also changed its apps from the [Lexus Link app](https://play.google.com/store/apps/details?id=app.mylexus.lexus.com.mylexus) to the [Lexus Link+](https://play.google.com/store/apps/details?id=com.lexus.oneapp.eu) app.
> 
> Which of the two apps do you use @Gorgounet?

Sorry I have been quiet as busy with work...
I use the Lexus Link + one!

**Reply by Gorgounet** [reply]
> I have a Lexus and I am using Lexus Link+
> 
> I have tried to execute the `simple_client_example.py`:
> 
> 
> 
> > ...\mytoyota-master\simple_client_example.py:86: DeprecationWarning: There is no current event loop
> 
> > loop = asyncio.get_event_loop()
> 
> > Logging in...
> 
> > Retrieving cars...
> 
> > [Finished in 2.7s]
> 
> 
> 
> I do not have any cars in MyToyota and I cannot add a Lexus car. I have to use Lexus Link+. 
> 
> 
> 
> I also found this and I have tried it:
> 
> ```
> 
> import requests
> 
> url = "https://ssoms.toyota-europe.com/authenticate"
> 
> payload = {'username': 'xx',
> 
>            'password': 'yy'}
> 
> files = []
> 
> headers = {
> 
>     'X-TME-BRAND': 'LEXUS',
> 
>     'X-TME-LC': 'ro-ro',
> 
>     'Accept': 'application / json, text / plain, * / *',
> 
>     'Sec-Fetch-Dest': 'empty'
> 
> }
> 
> response = requests.request(
> 
>     "POST", url, headers=headers, data=payload, files=files)
> 
> print(response.text)
> 
> ```
> 
> and I have modified  `X-TME-BRAND` from `TOYOTA` to `LEXUS`. I have the same response that I do not have a car (`"extras":{"hasPurchasedCars":false}`). If I will modify to something else (AUDI for example), I get this: 
> 
> 
> 
> > Brand is either not supported, missing or malformed
> 
> > [Finished in 804ms
> 
> 
> 
> Please let me know if you need additional information.
> 
> 

I think on that one in order to get a successful result you need to change the URL to something like https://ssoms.lexus.eu/authenticate

**Comment by CM000n** [comment]
Unfortunately, I can only help to a limited extent, as I don't own a Lexus myself. But it is not said that the existing methods will work 1:1 for you. You will probably have to play around with the end points yourself.
Perhaps @GitOldGrumpy has another useful tip?

**Reply by ellvisss** [reply]
Thank you anyway!
I am not so good at this and I am wondering if is possible to get the response with all the raw data, without "sorting" it. Maybe from there I can see something.

**Reply by CM000n** [reply]
What do you mean by "sorting"?

The endpoint tester basically does exactly that. It goes through all the specified endpoints and writes away the raw data of their response.

Lexus may use slightly different endpoints here. But for that we would have to take a look at their app and record the app traffic.

You will probably already fail at the authorization stage, as Lexus maybe uses a different URL here https://github.com/DurgNomis-drol/mytoyota/blob/f688783899dd760a88a322da43276f689970abc3/mytoyota/const.py#L46

**Reply by ellvisss** [reply]
I understood, sorry for the confusion.
I thought it will be something like a .json file with a lot of data (list, dictionary).

**Comment by GitOldGrumpy** [comment]
I suspect that it uses totally different endpoint Url's if you point me at the App you use I might be able to pull out the URL's so they can be tries.

**Reply by ellvisss** [reply]
This is the app that I am using: [Lexus Link+](https://play.google.com/store/apps/details?id=com.lexus.oneapp.eu&pcampaignid=web_share).

**Reply by Gorgounet** [reply]
I am happy to provide a logging with a car connected to it just for test purposes as I have access to multiple cars, I could create a logging with a car attached to it for you to play with

**Reply by GitOldGrumpy** [reply]
Investigations appear to show that the process for login to the Toyota and Lexus apps are the same, down to the same endpoint URL's used. The login process does what I expect and I can login in to the Lexus app using my Toyota credentials at which point the app does what I expect just receives no data as I don't own any Lexus cars. 

This probably means that there is a small difference I am yet to spot. If someone could:

- Edit `mytoyota\utils\logging\log_config.yaml` and change line 30 to `level: "DEBUG"`
- Run `simple_client_example.py` and provide all the output (Check for any personal identification information mainly email/password)

I can take a further look.

**Comment by ellvisss** [comment]
Here is the output: 
[output.txt](https://github.com/DurgNomis-drol/ha_toyota/files/14027486/output.txt)

Before editing the `mytoyota\utils\logging\log_config.yaml` and change line 30 to `level: "DEBUG"`, I tried to change to a wrong password and I would still get the same output as with the good password:

> C:\Temp\mytoyota-master\simple_client_example.py:88: DeprecationWarning: There is no current event loop
>   loop = asyncio.get_event_loop()
> Logging in...
> Retrieving cars...

Maybe I am doing something wrong?

**Reply by GitOldGrumpy** [reply]
@CM000n I think the logging output is broken?

**Reply by CM000n** [reply]
> @CM000n I think the logging output is broken?

Yeah, seems to be a problem with the current `stream` config. After deleting, logging to a file works again. https://github.com/DurgNomis-drol/mytoyota/pull/315

But the event loop error described by @ellvisss probably has more to do with the incompatibility of our asyncio example code in `simple_client_example.py` and the python env (python 3.11) he is using.:
https://stackoverflow.com/a/73367187

⚠️ **By the way, you shouldn't just share the whole log file with us in github @ellvisss!!! They may contain potentially privacy-sensitive information about you and your vehicle**

**Reply by GitOldGrumpy** [reply]
@ellvisss I have reviewed & merged @CM000n bug fix for logging. If you want to try changing the 3 lines at the bottom of simple_client_example.py:

From:

```
loop = asyncio.get_event_loop()
loop.run_until_complete(get_information())
loop.close()
```

To:

```
asyncio.run(get_information())
```

Then see how far we get, if you still get nothing do you want to capture the logs and provide. Thanks.

**Reply by ellvisss** [reply]
Thank you for reviewing it.
I have made the changes in `simple_client_example.py` and even with a good or wrong password, I still get the same output:
> Logging in...
> Retrieving cars...
Note: the error _DeprecationWarning: There is no current event loop   loop = asyncio.get_event_loop()_ is not there anymore.

This is the output after I changed level from `"ERROR"` to `"DEBUG"`:
good password: [output_2.txt](https://github.com/DurgNomis-drol/ha_toyota/files/14035115/output_2.txt)
wrong password: [output_wrong_password.txt](https://github.com/DurgNomis-drol/ha_toyota/files/14035147/output_wrong_password.txt)


Isn't it strange that with a wrong password, I do not have a login error?

**Reply by CM000n** [reply]
Please please, do not post any unchecked log files here 🙄 
At least I hope you looked at them beforehand?

Did you download the latest version of the mytoyota lib directly from the master branch? It looks like you are still using a version that does not include the fix from https://github.com/DurgNomis-drol/mytoyota/pull/315.

**Reply by ellvisss** [reply]
I did the changes requested in the [previous comment](https://github.com/DurgNomis-drol/ha_toyota/discussions/218#discussioncomment-8226151), I was not aware of the modifications from [DurgNomis-drol/mytoyota#315](https://github.com/DurgNomis-drol/mytoyota/pull/315)
I have looked at the logs and I mentioned what I could observe.. (as mentioned before, I am not that good at this).

I have replaced the `log_config.yaml` with the one from [DurgNomis-drol/mytoyota#315](https://github.com/DurgNomis-drol/mytoyota/pull/315)
The output is different now than before (with the level set to `DEBUG`) and it is similar with a wrong or good password for the Toyota/Lexus account: 
[output_3_good_pass.txt](https://github.com/DurgNomis-drol/ha_toyota/files/14035748/output_3_good_pass.txt)
[output_3_wrong_pass.txt](https://github.com/DurgNomis-drol/ha_toyota/files/14035750/output_3_wrong_pass.txt)

>  do not post any unchecked log files here 🙄

May I know please what do you expect to have more?

Thanks!

**Reply by CM000n** [reply]
> May I know please what do you expect to have more?

I don't want more, I want less 😉 
You should be aware that the logs may contain potentially privacy-sensitive information about you or your vehicles.
You should therefore only share specific parts of the logs that you are aware of, and not the complete, possibly unchecked, log files.

---

Now back to the topic. 😊 

I have seen that the header when retrieving the vehicles contains the app and brand in the form of `'x-channel': 'ONEAPP', 'x-brand': 'T'`.
This would probably have to be changed to the Lexus app and brand (possibly 'L').
You can try to play around with in in the `request_raw` function:
https://github.com/DurgNomis-drol/mytoyota/blob/master/mytoyota/controller.py#L238

**Reply by GitOldGrumpy** [reply]
So looking at the logs it has actually logged in, the second one with the wrong password works as it cache's the login tokens that are received from the first correct attempt. It then attempts to get the vehicles which works correctly **but** fails to find any.

As @CM000n suggests it could be that x-brand is set to 'T' not 'L' (which is correct for the Lexus app) but x-channel is 'ONEAPP' on both applications. I will check again this evening to see if I can see any more differances.

**Reply by ellvisss** [reply]
I tried to change the `x-brand` from `T` to `L`: [output_L.txt](https://github.com/DurgNomis-drol/ha_toyota/files/14040459/output_L.txt)

I also tried to change it to `AUDI` (to check if there is any difference, as this should not work) and it looks very similar: [output_AUDI.txt](https://github.com/DurgNomis-drol/ha_toyota/files/14040460/output_AUDI.txt)

**Reply by GitOldGrumpy** [reply]
I did have a suspicion it wouldn't work as I've noted before that it doesn't need to be set to 'T' for it to retrieve my cars from Toyota. I'm assuming that your cars turn up in the App on your phone?

I'll check for further differences this evening.

**Reply by ellvisss** [reply]
Yes, in the Lexus Link+ app the car is there and the data is updated.

**Comment by GitOldGrumpy** [comment]
The requests are identical apart from the brand, I mean so identical I have the same authorisation token in the Lexus app as I do in the Toyota app. It looks like its important to set all 3 of the headers for it too work. I say that as I can get it to break by setting them to "L" for my car. @ellvisss same line as before but instead of changing the line to "L" replace the single line with the following 3 lines:

```
"x-appbrand": "L",
"brand": "L",
"x-brand": "L",
```

It seems to default to "T" as I can remove it completely and it and it still works for me. I'm hopeful this makes it work or at least get further.

**Reply by GitOldGrumpy** [reply]
There maybe more issues as we only worked out what the minimum was required to get it to work for a Toyota, some of the extra headers that we don't add maybe required for the Lexus side. Log files if there any issues.

**Reply by ellvisss** [reply]
With those modifications we got further. I can have some information about the vehicle (looks more like the feature available). After that there were some errors: [output_4.txt](https://github.com/DurgNomis-drol/ha_toyota/files/14042723/output_4.txt)
I've changed the sensitive information with `PrivateInfo`

**Reply by CM000n** [reply]
This is a pydantic error, as the requirements for the model with vehicle information that we have created does not match the content of your vehicle.
Your vehicle does not provide the IMEI information that we would currently expect.

We will probably get a lot of errors like this for the different endpoints and have to make adjustments. But basically it seems to work. 😊 👍🏻

**Comment by GitOldGrumpy** [comment]
Ok, i've moved this to mytoyota [here](https://github.com/DurgNomis-drol/mytoyota/pull/316) as its not currently a home assistant issue. If we can get `simple_client_example.py` to work it should hopefully just work with the HA plugin.
